tensorflowonspark/pipeline.py

```diff
@@ -0,0 +1,443 @@
+    """
+    super(TFEstimator, self).__init__()
+    self.train_fn = train_fn
+    self.export_fn = export_fn
+    self.args = Namespace(tf_args) if isinstance(tf_args, dict) else tf_args
+    self.argv = tf_argv
+    self._setDefault(input_mapping={},
+                    cluster_size=1,
+                    num_ps=0,
+                    input_mode=TFCluster.InputMode.SPARK,
+                    protocol='grpc',
+                    tensorboard=False,
+                    model_dir=None,
+                    export_dir=None,
+                    tfrecord_dir=None,
+                    batch_size=100,
+                    epochs=1,
+                    readers=1,
+                    steps=1000)
+
+  def _fit(self, dataset):
+    """Trains a TensorFlow model and returns a TFModel instance with params/args pointing to a checkpoint/saved_model on disk.
+    Note: for input_mode == InputMode.TENSORFLOW, the dataset input is ignored, and it's assumed that the TF train_fn is responsible for reading input data.
+    """
+    sc = SparkContext.getOrCreate()
+
+    logging.info("===== 1. train args: {0}".format(self.args))
+    logging.info("===== 2. train params: {0}".format(self._paramMap))
+    local_args = self.merge_args_params()
+    logging.info("===== 3. train args + params: {0}".format(local_args))
+
+    if local_args.input_mode == TFCluster.InputMode.TENSORFLOW:
+      if dfutil.isLoadedDF(dataset):
+        # if just a DataFrame loaded from tfrecords, just point to original source path
+        logging.info("Loaded DataFrame of TFRecord.")
+        local_args.tfrecord_dir = dfutil.loadedDF[dataset]
+      else:
+        # otherwise, save as tfrecords and point to save path
+        assert local_args.tfrecord_dir, "Please specify --tfrecord_dir to export DataFrame to TFRecord."
+        df = dataset.select(self.getInputMapping().keys())
+        logging.info("Exporting DataFrame {} as TFRecord to: {}".format(df.dtypes, local_args.tfrecord_dir))
+        dfutil.saveAsTFRecords(df, local_args.tfrecord_dir)
+        logging.info("Done saving")
+
+    tf_args = self.argv if self.argv else local_args
+    cluster = TFCluster.run(sc, self.train_fn, tf_args, local_args.cluster_size, local_args.num_ps, local_args.tensorboard, local_args.input_mode)
+    if local_args.input_mode == TFCluster.InputMode.SPARK:
+      # feed data, using a deterministic order for input columns (lexicographic by key)
+      input_cols = sorted(self.getInputMapping().keys())
+      cluster.train(dataset.select(input_cols).rdd, local_args.epochs)
+    cluster.shutdown()
+
+    # Run export function, if provided
+    if self.export_fn:
+      assert local_args.export_dir, "Export function requires --export_dir to be set"
+      logging.info("Exporting saved_model (via export_fn) to: {}".format(local_args.export_dir))
+
+      def _export(iterator, fn, args, argv=None):
+        single_node_env(args)
```

Comment:
I notice that sample code include "sys.argv = argv". Should that be set by single_node_env()?